modules/fedimint-ln-client/src/pay.rs

```diff
@@ -0,0 +1,548 @@
+///  Refund -- await transaction rejected --> Failure
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum LightningPayStates {
+    CreatedOutgoingLnContract(LightningPayCreatedOutgoingLnContract),
+    Canceled,
+    Funded(LightningPayFunded),
+    Success(String),
+    Refundable(LightningPayRefundable),
+    Refund(LightningPayRefund),
+    Refunded(TransactionId),
+    Failure(TransactionId),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct LightningPayCommon {
+    pub operation_id: OperationId,
+    pub federation_id: FederationId,
+    pub contract: OutgoingContractData,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct LightningPayStateMachine {
+    pub common: LightningPayCommon,
+    pub state: LightningPayStates,
+}
+
+impl State for LightningPayStateMachine {
+    type ModuleContext = LightningClientContext;
+    type GlobalContext = DynGlobalClientContext;
+
+    fn transitions(
+        &self,
+        _context: &Self::ModuleContext,
+        global_context: &Self::GlobalContext,
+    ) -> Vec<StateTransition<Self>> {
+        match &self.state {
+            LightningPayStates::CreatedOutgoingLnContract(created_outgoing_ln_contract) => {
+                created_outgoing_ln_contract.transitions(&self.common, global_context)
+            }
+            LightningPayStates::Canceled => {
+                vec![]
+            }
+            LightningPayStates::Funded(funded) => funded.transitions(),
+            LightningPayStates::Success(_) => {
+                vec![]
+            }
+            LightningPayStates::Refundable(refundable) => {
+                refundable.transitions(self.common.clone(), global_context.clone())
+            }
+            LightningPayStates::Refund(refund) => refund.transitions(&self.common, global_context),
+            LightningPayStates::Refunded(_) => {
+                vec![]
+            }
+            LightningPayStates::Failure(_) => {
+                vec![]
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        self.common.operation_id
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct LightningPayCreatedOutgoingLnContract {
+    pub funding_txid: TransactionId,
+    pub contract_id: ContractId,
+    pub gateway: LightningGateway,
+}
+
+impl LightningPayCreatedOutgoingLnContract {
+    fn transitions(
+        &self,
+        common: &LightningPayCommon,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<LightningPayStateMachine>> {
+        let txid = self.funding_txid;
+        let contract_id = self.contract_id.clone();
+        let funded_common = common.clone();
+        let rejected_common = funded_common.clone();
+        let success_context = global_context.clone();
+        let reject_context = global_context.clone();
+        let gateway = self.gateway.clone();
```

Comment:
For simplicity's sake while implementing this I just cloned everything. I don't think that's too expensive but its probably not necessary.